esp/services/ldapenvironment/ldapenvironmentService.cpp

```diff
@@ -0,0 +1,204 @@
+            throw MakeStringException(-1, "hpccRootName must be specified in configuration (ex. 'hpcc,dc=myLDAPDomain,dc=com')");
+
+        xpath.clear().appendf("%sadminGroupName", prefix.str());
+        cfg->getProp(xpath.str(), adminGroupName);
+        if (adminGroupName.isEmpty())
+            adminGroupName.set("HPCCAdmins");
+
+        xpath.clear().appendf("%ssharedFilesBaseDN", prefix.str());
+        cfg->getProp(xpath.str(), sharedFilesBaseDN);
+        if (sharedFilesBaseDN.isEmpty())
+            throw MakeStringException(-1, "sharedFilesBaseDN must be specified in configuration (ex. 'ou=files,ou=shared,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+        xpath.clear().appendf("%ssharedGroupsBaseDN", prefix.str());
+        cfg->getProp(xpath.str(), sharedGroupsBaseDN);
+        if (sharedGroupsBaseDN.isEmpty())
+            throw MakeStringException(-1, "sharedGroupsBaseDN must be specified in configuration (ex. 'ou=groups,ou=shared,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+        xpath.clear().appendf("%ssharedUsersBaseDN", prefix.str());
+        cfg->getProp(xpath.str(), sharedUsersBaseDN);
+        if (sharedUsersBaseDN.isEmpty())
+            throw MakeStringException(-1, "sharedUsersBaseDN must be specified in configuration (ex. 'ou=users,ou=shared,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+        xpath.clear().appendf("%ssharedResourcesBaseDN", prefix.str());
+        cfg->getProp(xpath.str(), sharedResourcesBaseDN);
+        if (sharedResourcesBaseDN.isEmpty())
+            throw MakeStringException(-1, "sharedResourcesBaseDN must be specified in configuration (ex. 'ou=smc,ou=espservices,ou=shared,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+        xpath.clear().appendf("%ssharedWorkunitsBaseDN", prefix.str());
+        cfg->getProp(xpath.str(), sharedWorkunitsBaseDN);
+        if (sharedWorkunitsBaseDN.isEmpty())
+            throw MakeStringException(-1, "sharedWorkunitsBaseDN must be specified in configuration (ex. 'ou=workunits,ou=shared,ou=hpcc,dc=myLDAPDomain,dc=com')");
+
+    }
+    catch (...)
+    {
+        throw MakeStringException(-1, "ldapenvironment: Error querying environment settings");
+    }
+}
+
+bool CldapenvironmentEx::onLDAPQueryDefaults(IEspContext &context, IEspLDAPQueryDefaultsRequest &req, IEspLDAPQueryDefaultsResponse &resp)
+{
+    resp.setHPCCRootName(ldapRootOU.str());
+    resp.setAdminGroupName(adminGroupName.str());
+    resp.setSharedFilesBaseDN(sharedFilesBaseDN.str());
+    resp.setSharedGroupsBaseDN(sharedGroupsBaseDN.str());
+    resp.setSharedUsersBaseDN(sharedUsersBaseDN.str());
+    resp.setSharedResourcesBaseDN(sharedResourcesBaseDN.str());
+    resp.setSharedWorkunitsBaseDN(sharedWorkunitsBaseDN.str());
+    resp.setRetcode(0);
+    return true;
+}
+
+const char * CldapenvironmentEx::formatOUname(StringBuffer &ou, const char * envName, int mode, const char * sharedOU, const char * reqBaseDN, const char * privateOU)
+{
+    switch (mode)
+    {
+        case 0: // Generated Private OU
```

Comment:
Use the values generated from the ESDL in the ECM file:

case COUMode_GeneratePrivate:
case COUMode_UseGlobal:
case COUMode_CreateCustom:

The ESDL creates a c++ enum:

`enum COUMode { OUMode_Undefined=-1,COUMode_GeneratePrivate=0, COUMode_UseGlobal=1, COUMode_CreateCustom=2, };
`

Reply:
Cool, added to various places in my code